### ISSUE TITLE
Support for fixity verification?

### ISSUE THREAD

**Issue by awoods** [fcrepo/fcrepo-specification#373]
There is interest from Fedora stakeholders to have a "fixity checking service" that:
- calculates the current checksum of a binary resource
- compares the calculated checksum with a stored value
- responds with `success` or `failure` based on the comparison.

The question here is whether the `Want-Digest` header could invoke such a service? The limitations or expectations on responses to `Want-Digest` requests are unclear in the reading of [rfc-3230](https://tools.ietf.org/html/rfc3230#section-4.3.1).

**Comment by barmintor** [comment]
I feel like RFC3230 reads pretty straightforwardly: `Want-Digest` requests that the response include a `Digest` header in the algorithm specified by `Want-Digest`. The scope of the `Digest` header's value is the instance digest of the resource identified by the request URI. The message body does not have to be the instance, in whole or in part - this allows for HEAD or partial responses. What is the ambiguous read?

**Comment by barmintor** [comment]
The ambiguities I see are not relevant to what's being proposed here:
1. You can include `contentMD5` as an algorithm to indicate that `Content-MD5:` header should be included in the response, but `Content-MD5` is a message digest header and not an instance digest header.

2. There's nothing specifically addressing when the value of `Digest:` was calculated.

**Comment by awoods** [comment]
I am not seeing where rfc-3230 says: "`Want-Digest` requests that the response include a `Digest` header".

...although that may be inferred from: 
> "The Want-Digest message header field indicates the sender's desire to receive an instance digest"

Of course, rfc-3230 takes no position on responding with additional headers as well.

**Comment by barmintor** [comment]
That's not _false_, but treating the two headers as unrelated in the spec that introduces them *and* the term "instance digest" seems like a willful misreading, doesn't it?

**Comment by zimeon** [comment]
I feel uncomfortable attempting to interpret `Want-Digest` as something other than "I want a digest back". I also would not want remove this clear and simple case that supports a client making a digest check.

It seems to me that "I want the server to check the digest" is a different and new feature. At present there is nothing in the specification the says the implementation has to store a digest -- a prerequisite for doing an internal digest check. Thus there is the possibility of a third possible response: "I can't".  I think we might be best off creating a new header (maybe `Check-Digest`) in order to request this internal checking on POST (because adding a new HTTP method is too bold). Response 200 + header values for match / no-match, response 501 for not supported.

(Another idea, maybe crazy, would be to think of this as a precondition and invent an `If-Digest-Match` header that behaves similarly to the `If-Match` header. So 2xx if OK, 412 if bad, 501 if not supported.)

**Comment by awoods** [comment]
Moving this conversation away from `Want-Digest` seems appropriate. As also suggested by @dshalvi on today's: https://wiki.duraspace.org/display/FF/2018-05-24+-+Fedora+Tech+Meeting

**Comment by dannylamb** [comment]
@zimeon I've had the same crazy thoughts

**Comment by barmintor** [comment]
@zimeon a depressingly long time ago we were bandying around `Expect:` ideas until @birkland pointed out that it broke caches and that the HTTP community had basically abandoned the idea except for `100-continue`.

**Comment by kefo** [comment]
Why a `POST`?  Perhaps it shouldn't, but that makes me nervous.  I would have expected such a service to respond to a `GET` or `HEAD` request.  

I imagine the permissible header values (such as sha, sha256, md5) would be only the digest algorithms supported by the repository?

Anyways, `Want-Digest` is a different request from `Check-Digest` or `If-Digest-Match` or some other variant, so the issue's name is a little misleading.

**Comment by awoods** [comment]
Renamed title of this issue from: "Want-Digest to verify fixity?" to "Support for fixity verification?"

**Comment by escowles** [comment]
@kefo I don't think any of the standard HTTP verbs are a good fit here.  The core of a fixity check is requesting that an action take place (perform a checksum), which seems like a bad fit for `GET`.  In particular, making a `GET`/`HEAD` that is explicitly non-cacheable seems like a bad pattern.

Though it's not a great fit for `POST` either, and nothing else seems better.

**Comment by dannylamb** [comment]
Yeah, because we're really talking about a service and not a resource here.  IMO there is no "good" verb to do this sort of thing if you still want to request against the resource.  For me, POST ekes out the winner because GET would have to pass in Cache-Control headers to explicitly disable caching.  We'll also have to be explicit that no resource will get created if the `If-Digest-Match` header is present.

**Comment by barmintor** [comment]
If you're talking about `If-Digest-Match`, you could also pivot back to `Expect: 100-continue` with GET. The possibilities are vast if you allow 2 HTTP requests and invent a request header.

**Comment by barmintor** [comment]
I think that it's ok for an action to take place - providing a representation is an action - the issue is one of resource mutability/idempotency as far as I understand the RESTful semantics of GET.

edit: This is the same reason that early attempts at this spec realized the digest check as a resource - if the result of the check has a representation, the GET/POST semantics are clearer.

**Comment by dannylamb** [comment]
@barmintor So the idea is, send GET with `Expect: 100-continue` and `If-Digest-Match: xxxxxxx`.  If you get 100, fixity check succeeds.  412 it fails, and 501 means the implementation does not support it.  Sound about right?

**Comment by barmintor** [comment]
I don't see a way towards the 501 part of that proposal - not supporting it is something that would have to be inferred from a response that was neither 100 nor 417. But I also feel like we hashed a lot of this out with @birkland and went away from it for a reason re: proxies. Inventing a header might work better as `If-None-Digest`, with a digest token, and having anything but a 304 mean the check failed or was unsupported. But if you want to distinguish a failure from a missing implementation, you... send Want-Digest and see if matches what you also sent in `If-None-Digest`, and we're back where we started.

**Comment by dannylamb** [comment]
Am I interpreting that pattern correctly?  Because I'm ok with that and would rather not conflate services like this with resource creation and POST.

**Comment by dannylamb** [comment]
haha... should've cleared my cache.  reading your response now @barmintor

**Comment by kefo** [comment]
@dannylamb @escowles I don't want to belabor this, but I think it is worth noting two things with respect to GET/HEAD:

1) `Want-Digest` could also be seen as a type of service, in that you are asking the server to perform an action, that is generate the digest (presumably) on demand.
2) There is already a [non-normative note](https://fcrepo.github.io/fcrepo-specification/#http-get-ldpnr) advising clients to include a `Cache-Control: no-cache` header in the request for `Want-Digest` precisely because of the caching issue.

It seems the pattern we're discussing here (perform a fixity check based on a request header and include the result of said check in a response header) is fairly close to the pattern defined for `Want-Digest` (calculate the digest based on a request header and include the result in a response header).

FWIW, I foresee clients performing persistence fixity checking to only make `HEAD` requests.  If a client were to perform a `GET` and also invoke the fixity check service I would think it perfectly reasonable that the client receive the binary as the response body, just as the client would now (per the spec) if it included a `Want-Digest` header in a GET request.

**Comment by dannylamb** [comment]
@kefo I can get to that.  A GET/HEAD request with `If-Digest-Match: XXXX` and `Cache Control: no-cache` with 200 and the returned entity on success, 412 on fail, and some other 4xx or 5xx for "can't/won't do it".

Laid out like that it seems straightforward enough for a situation like this.  In the least it's in line with what we're doing elsewhere.

**Comment by escowles** [comment]
Another option I've thought about before is having a fixity endpoint that is more similar to the Memento pattern, with a container for holding fixity results.  The interactions would look something like:

* In a LDP-NR's headers: `Link: <path/to/fixity/endpoint>; rel="fixity"`
* The fixity endpoint would be a LDPC that had a triple specifying the predicate used for determining the stored checksum for the LDP-NR.  Implementations MAY allow PATCH on the LDPC to alter which predicates are used
* POST to the LDPC with empty body would calculate the current checksum, compare it to the stored checksum, and create an LDP-RS summary of the fixity check, which could be deleted, but not modified
* POST to the LDPC would generate a message

I would be 👍 to specifying something like this in a sidecar spec.

**Comment by peichman-umd** [comment]
@escowles I like your suggestion of a Memento-like container pattern.

Two questions:

1. When you have the two bullet points starting with "POST to the LDPC", is that intended to list 2 results of a single POST request? (i.e., calculate checksum and create an LDP-RS summary, and emit an ActivityStreams message on the JMS bus)
2. What are the advantages of putting this in a sidecar spec over leaving it in the core Fedora API spec?

**Comment by awoods** [comment]
@peichman-umd : I believe the value in a sidecar spec is that it not interfere with getting the Fedora Specification 1.0 released (and associated implementations).
The sidecar effort would develop in parallel to the core spec... and ideally roll into the core spec in release 1.1.

**Comment by escowles** [comment]
There is an existing draft from earlier fixity discussions, that outlines a Memento-style fixity approach:
https://docs.google.com/document/d/1ws1x7bry3RqB305XxtqIgru5c8heTDDk03jkbCA7iW4/edit

@peichman-umd I agree with @awoods' assessment: the main benefit of a sidecar spec is that it doesn't block the 1.0 release of the spec.  Since 1.0 is waiting on implementations at this point, it seems counter-productive to add new requirements that would make that more work.  So a sidecar spec seems like the best way to let this work continue, be implemented, and then be folded into the spec in a post-1.0 release.

**Comment by peichman-umd** [comment]
@escowles @awoods I am fine with this being in a sidecar spec, with the proviso that we keep on top of it as a community and don't let it languish 😄

**Comment by awoods** [comment]
Deferring until v2 of specification per: https://wiki.duraspace.org/display/FF/2018-08-29+-+API+Spec+Meeting

**Comment by ioggstream** [comment]
@awoods we're rewriting rfc-3230 specs and probably deprecate md5. Consider switching to sha256 if you're using digest.

This thread is very interesting though: feel free to contribute to the above issue ^.

See https://httpwg.org/http-extensions/draft-ietf-httpbis-digest-headers.html
